### PR TITLE
Use displacement instead of initial bucket in HashMap code

### PR DESCRIPTION
**Pull request by arthurprs**
Use displacement instead of initial bucket in HashMap code. It makes the code a bit cleaner and also saves a few instructions (handy since it'll be using some to do some sort of adaptive behavior soon).

### CODE REVIEW

**Comment by arthurprs** [comment]
I botched the patch as always, will fix in the morning.

File: src/libstd/collections/hash/map.rs [hunk 1]
```diff
@@ -522,11 +523,11 @@ impl<K, V, S> HashMap<K, V, S>
 
     // The caller should ensure that invariants by Robin Hood Hashing hold.
     fn insert_hashed_ordered(&mut self, hash: SafeHash, k: K, v: V) {
-        let raw_cap = self.raw_capacity();
+        let capacity = self.table.capacity();
```
Comment (bluss) [inline src/libstd/collections/hash/map.rs]: this is still the raw capacity (change is equivalent, so seems unnecessary to change).

File: src/libstd/collections/hash/map.rs [hunk 2]
```diff
@@ -537,8 +538,8 @@ impl<K, V, S> HashMap<K, V, S>
                 Full(b) => b.into_bucket(),
             };
             buckets.next();
+            debug_assert!(buckets.index() < limit_bucket);
```
Comment (bluss) [inline src/libstd/collections/hash/map.rs]: This is an infinite loop if there is no free bucket, but that there is, is a requirement already in the caller (resize), so it seems ok.

(Note to other readers: buckets.index() keeps increasing even if the pointer it corresponds to wraps back to the first bucket.)

**Review by bluss — CHANGES_REQUESTED** [review]
Looks good to me, there's just a comment that needs to be fully updated.

**Comment by arthurprs** [comment]
Thanks for the review. Commit amended with fixed/new comments too.

**Review by bluss — APPROVED** [review]

**Comment by bluss** [comment]
@bors r+

Thanks

**Pull request merged**